### PR TITLE
scheduler: support the hot-region-scheduler to generate multiple operators at the same time

### PR DESCRIPTION
**Pull request by HunDunDM**
### What problem does this PR solve?

Issue Number: Ref #4949 

### What is changed and how does it work?

Allows a solution to contain multiple Operators to expand the search set, guaranteeing that a better solution can be found in some cases.

### Check List

Tests

- Unit test

### Release note

```release-note
Optimized the search strategy of the balance-hot-region-scheduler.
```

### CODE REVIEW

**Comment by nolouch** [comment]
ptal @lhy1024

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -497,8 +537,43 @@ func (bs *balanceSolver) solve() []*operator.Operator {
 				bs.cur.dstStore = dstStore
 				bs.calcProgressiveRank()
 				tryUpdateBestSolution()
+
+				if searchRevertRegions && (bs.cur.progressiveRank >= -1 && bs.cur.progressiveRank <= 0) &&
```
[the remark this answers is not in the document]
Comment (lhy1024) [reply server/schedulers/hot_region.go:565]: I think we should support rank1 later.

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -521,12 +561,52 @@ func (bs *balanceSolver) solve() []*operator.Operator {
...
 }
 
+func (bs *balanceSolver) allowSearchRevertRegions() bool {
+	// The next solve is allowed to search-revert-regions only when the following conditions are met.
+	// * No best solution was found this time.
+	// * The progressiveRank of the best solution is -1.
+	// * The best solution contain revert regions.
+	return bs.best == nil || bs.best.progressiveRank >= -1 || len(bs.best.revertRegions) > 0
+}
+
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:609]: For the multi operator, pending should be the result of taking the absolute value.

File: server/schedulers/hot_region.go [hunk 3]
```diff
@@ -1094,12 +1194,24 @@ func (bs *balanceSolver) buildOperators() (ops []*operator.Operator, infl *stati
...
+			currentOp, typ, err = createOperator(revertRegion, dstStoreID, srcStoreID)
+			if err != nil {
+				break
+			}
+			bs.decorateOperator(currentOp, true, targetLabel, sourceLabel, typ, dim)
+			ops = append(ops, currentOp)
+			for j, load := range bs.cur.dstPeersStat[i].Loads {
+				infl.Loads[j] += load
+			}
+			infl.Count++
```
Comment (bufferflies) [inline server/schedulers/hot_region.go:1213]: should we need keep the influence disappear by atomicity?

**Review by lhy1024 — APPROVED** [review]
When the resources are sufficient, the two branches of master and multiop behave similarly, and the solution can be found (split.qps-threshold=500) and there is almost no scheduling after stabilization, and there is still scheduling after the solution cannot be found (split.qps-threshold=3000) ; When resources are insufficient, the number of peak scheduling of multiop is more

File: server/schedulers/hot_region.go [hunk 4]
```diff
@@ -1095,18 +1207,22 @@ func (bs *balanceSolver) buildOperators() (ops []*operator.Operator, infl *stati
...
-		infl = &statistics.Influence{
-			Loads: append(bs.cur.srcPeerStat.Loads[:0:0], bs.cur.srcPeerStat.Loads...),
-			Count: 1,
+		for _, revertRegion := range bs.cur.revertRegions {
+			currentOp, typ, err = createOperator(revertRegion, dstStoreID, srcStoreID)
+			if err != nil {
+				break
+			}
+			bs.decorateOperator(currentOp, true, targetLabel, sourceLabel, typ, dim)
+			ops = append(ops, currentOp)
```
Comment (nolouch) [inline server/schedulers/hot_region.go:1218]: I wonder if the `RandBuckets` in`waiting_operator` will be influenced. The `GetOperator` may only get first one operator and miss the second one.

File: server/schedulers/hot_region.go [hunk 5]
```diff
@@ -329,10 +331,15 @@ func (h *hotScheduler) balanceHotWriteRegions(cluster schedule.Cluster) []*opera
...
-	srcPeerStat *statistics.HotPeerStat
-	region      *core.RegionInfo
-	dstStore    *statistics.StoreLoadDetail
+	srcStore     *statistics.StoreLoadDetail
+	region       *core.RegionInfo // The region of the main balance effect. Relate mainPeerStat. srcStore -> dstStore
+	mainPeerStat *statistics.HotPeerStat
+
+	dstStore        *statistics.StoreLoadDetail
+	revertRegions   []*core.RegionInfo // The regions to hedge back effects. Relate revertPeersStat. dstStore -> srcStore
+	revertPeersStat []*statistics.HotPeerStat
```
Comment (nolouch) [inline server/schedulers/hot_region.go:340]:
```suggestion
	revertPeersStat *statistics.HotPeerStat
```
I prefer it like this. if you want to support more operators, may also need to make `mainPeerStat` to `[]*statistics.HotPeerStat` in the future.

**Review by nolouch — COMMENTED** [review]
rest LGTM

**Review by lhy1024 — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**